{% extends "base.html" %}

{% block content %}
<div class="container container-main" role="main">
    <h2>Help</h2>

    <!--TODO: take updated screenshots... -->

    <h3 id="uploading_datasets">Uploading datasets</h3>

    <figure class="figure">
        <img src="{{ url_for('static', filename='upload_1.png') }}" style="width: 600px;" class="figure-img center-block" alt="upload screen">
        <figcaption class="figure-caption text-center">Upload screen</figcaption>
    </figure>
    </br>

    <p>To upload a new dataset, select a data file containing a gene expression matrix. This file can be a sparse matrix file in Matrix Market format (.mtx, as produced by scipy.io.mmwrite) or a dense matrix as a space or tab-delimited file (as produced by np.savetxt). The file should not contain any headers or column names (e.g. gene names or barcodes). Make sure to indicate whether the file is genes-by-cells or cells-by-genes. The input file may be gzipped.</p>

    <p>Multiple datasets can be uploaded together; these datasets will be combined. In order to upload multiple datasets to be combined, click on the TODO</p>

    <h3 id="running_uncurl">Running uncurl</h3>

        <img src="{{ url_for('static', filename='data_preview.png') }}" style="width: 600px;" class="figure-img center-block" alt="Data preview">
        <figcaption class="figure-caption text-center">Data preview</figcaption>
    </figure>
    </br>

    <p>After uploading the data, you will eventually be redirected to a view that looks like the one above. On the top, there are two plots. The plot on the left shows the distribution of total read counts per cell. The plot on the right shows the relationship between mean and variance for all genes.</p>

    <h4>Uncurl options</h4>

    <figure class="figure">
        <img src="{{ url_for('static', filename='uncurl_params.png') }}" style="width: 500px;" class="figure-img center-block" alt="Uncurl parameters">
        <figcaption class="figure-caption text-center">Uncurl parameters</figcaption>
    </figure>
    </br>

    <ul>
        <li>Number of clusters: If the number of cell types is known, use the number of cell types here. Otherwise, this can be set to 0, which automatically infers the number of clusters using a gap score metric. Alternatively, setting this to between 10 and 20 usually produces reasonable results (clusters can be merged or split later on).</li>
        <li>Baseline visualization method: How to create a scatterplot of cells without using Uncurl. One of PCA, tSNE, or UMAP. Default: tSNE.</li>
        <li>Visualization method: How to create a scatterplot of cells using the output of Uncurl. One of MDS, PCA, tSNE, or UMAP. Default: tSNE.</li>
        <li>Remove cells with read counts greater/less than: This should be self-explanatory. Cells removed this way will not be used in the downstream processing. The default is to remove the top and bottom 5% of cells by total read count. If you want to include all cells, set the first number to 0 and the second number to a very high number.</li>
        <li>Distribution: For UMI or count-valued data, Poisson usually is best. For normalized counts (FPKM, RPKM,...), you might want to use log-normal. If the data has already been log-transformed, Gaussian might be preferable. Default: Poisson.</li>
        <li>Fraction of genes to include: As a preprocessing step, Uncurl selects a subset of genes for inference using variance-based selection: binning the data into 5 bins by mean expression level, and selecting the top <i>k</i> fraction of genes by variance. By default, this is 0.2.</li>
        <li>Fraction of cells to include for visualization: If there is a large number of cells, then the number of cells can be subsampled for the dimensionality reduction, to reduce computational and data transfer requirements. By default, this sets it so that the number of cells retained is at most about 10000. Set this to 1.0 to include all cells.</li>
        <li>Normalize cells by read count: If this is set to true, then for all cells, the read counts will be divided by the total read count for that cell, and then multiplied by the median read count for all cells. Default: true.</li>
    </ul>

    <h3 id="user_interface">User interface</h3>

    <figure class="figure">
        <img src="{{ url_for('static', filename='scatterplot_1.png') }}" style="width: 900px;" class="figure-img center-block" alt="Scatterplot + barplot">
        <figcaption class="figure-caption text-center">Visualization: scatterplot of means + barplot</figcaption>
    </figure>
    </br>

    <p>After UNCURL has finished running, you will be redirected to a page that looks like the one above. The graph on the left is a scatterplot that shows a dimensionally reduced view of the cells or cluster means. The graph on the right shows the top genes for the selected cluster, or relationships between clusters.</p>

    <p>To change the scatterplot view, click on the radio buttons above the plot, circled in red. The default view shows a scatterplot of all of the cells in the uploaded dataset.</p>

    <p>To change the cluster being shown on the barplot, click on any cell or cluster mean.</p>

    <p>To change the color scheme, use the "Label scheme" dropdown. This also allows you to upload a new color scheme for visualization.</p>

    <h3>Interacting with the plot</h3>

    <p>Double click on a cluster name to see only that cluster. Single click on a cluster name in the legend to toggle its visibility.</p>

    <p>Mousing over the top right of the plot will show a control panel. This allows you to zoom in/out, select cells, or save the plot.</p>

    </hr>

    <h3>Scatterplot views</h3>

    <!-- TODO: describe all of the different views present in the scatterplot right now. -->
    <p>You can select different views for the main scatterplot by clicking on the radio buttons above. There are a number of different scatterplot views:</p>

    <ul>
        <li><b>Pre-processed cells</b> - This is the default view, which shows each cell as a dot on a 2D plane. The underlying data comes from a tSNE or other dimensionality reduction method.</li>
        <li><b>Post-processed cells</b> - This is the same as the previous, but uses the post-UNCURL data as input into the dimensionality reduction.</li>
        <li><b>Cluster means</b> - This shows the mean position of each cluster as a point.</li>
        <li><b>Label heatmap</b> - This view shows the relationship between two different label sets. The colors represent the overlap between cells.</li>
        <li><b>Dendrogram heatmap</b> - This shows a dendrogram of the selected genes for the clusters.</li>
        <li><b>Gene</b> - This shows each gene as a dot. The underlying data comes from a umap treating each gene as a data point.</li>
        <li><b>Gene heatmap</b> - This shows the correlations among a set of genes.</li>
    </ul>

    <h3>Barplot views</h3>

    <!-- TODO: describe all the different views present in the barplot right now. -->
    <p>You can select different views for the barplot by selecting them in the dropdown on the top right of the page. There are a number of different barplot views:</p>

    <ul>
        <li><b>Top genes (1-vs-rest fold change)</b> - The default view. This shows the genes for the selected cluster that have the highest fold change compared to the rest of the cells. Click on a cell on the scatterplot, and the corresponding cluster will be shown here.</li>
        <li><b>Top genes (p-value of fold change)</b> - This shows the genes for the selected cluster that have the lowest p-value of the fold change, based on a t-test. Again, clicking on a cell from the scatterplot will show the results for the cell's cluster here.</li>
        <li><b>Pairwise fold change</b> - This allows the user to test the differential expression between two selected clusters, showing the genes with the highest fold-change between cluster 1 and cluster 2.</li>
        <li><b>Pairwise p-value</b> - Same as above, but for the genes with the lowest p-value of the t-test.</li>
    </ul>

    <h3>Cell labels</h3>

    <p>The cell labels dropdown, below the scatterplot, allows the user to change the color scheme used for the scatterplot. If a set of categorical cell labels is uploaded or created, UNCURL-App will automatically calculate differential expression for those cell labels, which will be shown on the barplot. Here are the options for cell labels:</p>

    <ul>
        <li><b>Cluster</b> - colors the cells based on the inferred clusters.</li>
        <li><b>Entropy</b> - colors the cell based on</li>
        <li><b>Gene</b> - colors the cells based on the expression level for a specified gene or genes (sum of expression, input as a space-separated list).</li>
        <li><b>Cluster weights</b> - colors the cells based</li>
        <li><b>Read counts</b> - colors the cells based on read counts.</li>
        <li><b>Upload cell labels</b> - This allows the user to upload custom cell labels. The uploaded file should have a line for every cell uploaded, and each line should have a cluster name or ID. After uploading the cell labels, this dropdown will be updated to include the name of the uploaded file as one of the options.</li>
        <li><b>Custom cell selection</b> - This allows the user to create custom cell selections based on various features. The user may select cells based on combinations of existing clusters, gene expression, or uploaded cell labels.</li>
    </ul>

    </hr>
    
    <h3>Gene set databases</h3>

    <p>UNCURL-App currently contains interfaces to three gene set databases: Enrichr, CellMarker, and CellMeSH. These databases can be used to aid in identifying cell types corresponding to clusters.</p>

    <h4>CellMeSH</h4>

    <p>CellMeSH is the default database that can be used to help identify the cell type of a given cluster. CellMeSH currently includes human and mouse genes/cell types.</p>

    <p>To use CellMeSH to identify the cell type of a cluster:</p>
    
    <ol>
        <li>Click on a cell on the scatterplot. The set of genes used in the cell type query are the same genes that are shown on the barplot. Changing the barplot view will change the genes that are used for querying.</li>
        <li>On the very bottom of the screen, click "Submit CellMesH query". Then, a list of cell types will be shown, ordered by descending similarity to the query.</li>
    </ol>

    <p>Other gene set databases can be queried in the same way as CellMesh.</p>

    <h4>Enrichr</h4>
    
    <p>This is an interface to the Enrichr tool (http://amp.pharm.mssm.edu/Enrichr/). This does not include all gene set libraries present in Enrichr, just the ones that might be helpful in identifying cell types.</p>

    <h4>CellMarker</h4>

    <p>A copy of the <a href="http://biocc.hrbmu.edu.cn/CellMarker/">CellMarker</a> database can be used for cell type identification/gene set querying.</p>

    <h4>Gene Ontology</h4>

    <p></p>

    <h4>KEGG</h4>

    <p></p>


    <!-- TODO: insert picture here -->

    </hr>

    <h3>Merging, splitting, re-analysis</h3>

    <p>Using the scatterplot, the user can merge or split existing clusters, or create a new cluster from selected cells.</p>

    <p>To <b>merge</b> multiple clusters:</p>

    <ol>
        <li>Mouse over to the top right corner of the screen, and select either the "Box Select" or "Lasso Select" tool.</li>
        <li>Draw a border around the clusters that should be merged. You should see a text box below the scatterplot that mentions the selected clusters.</li>
        <li>Click on the "Recluster" button below the scatterplot, and then click "Merge clusters".</li>
        <li>Wait for the process to finish. It might take around the same time as the original data upload.</li>
    </ol>

    <p>To <b>split</b> a cluster:</p>
    <ol>
        <li>Click on a cell belonging to the cluster that should be split.</li>
        <li>Click on the "Recluster" button, and then click "Split clusters".</li>
        <li>Wait for the process to finish. It might take around the same time as the original data upload.</li>
    </ol>

    <p>To <b>delete</b> a group of cells or a cluster:</p>

    <p>To assign a group of cells to a new cluster:</p>


    <h3>Experimental Features</h3>

    <p>These features are not supported.</p>

</div>

{% endblock %}
