{% extends "base.html" %}

{% block head %}

<!-- TODO: have an option for serving static files?-->

<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js"></script>

{{ super() }}

<script src="https://code.jquery.com/jquery-3.1.0.min.js"></script>
<script src="https://cdn.plot.ly/plotly-latest.min.js"></script>

<script>
// this script is used to generate all the graphs, and deal with callbacks. 
var user_id = "{{ user_id }}";

// have a user-side cache to avoid having to re-make so many json calls
var cache = {};
cache.barplots = {};
cache.scatterplots = {};
cache.enrichr = {};
cache.cellmarker = {};

var currently_selected_cluster = 0;

var current_scatterplot_data = {};

var all_selected_clusters = [];
var current_selected_cells = [];
var currently_merging = false;

// use jquery for ajax calls

function bind_click() {
    var plot = $('#means-scatter-plot')[0];
    plot.on('plotly_click', function(data) {
        var cluster = data.points[0].curveNumber;
        current_selected_cells = [];
        for (var i = 0; i<data.points.length; i++) {
            current_selected_cells.push(data.points[i].text);
        }
        // update selected cluster
        currently_selected_cluster = cluster;
        var top_or_bulk = $("#top-or-bulk").val();
        // set cluster 1 to be the selected value
        if (top_or_bulk.includes('pairwise')) {
            $('#barplot_cluster_select_1').val(String(data.points[0].curveNumber));
        }
        console.log(cluster);
        update_barplot(cluster);
        // update barplot
        var l = current_scatterplot_data.data[cluster].x.length;
        all_selected_clusters = [cluster];
        selection_string = "cluster " + String(cluster) + 
            " (" + String(l) + " cells), ";
        // set update area...
        $("#update-area").empty();
        $("#update-area").append("selected clusters: " + selection_string);
    });
};

// function called whenever a scatterplot is clicked or
// when the dropdown is changed...
function update_barplot(cluster_number) {
    var top_or_bulk = $("#top-or-bulk").val();
    var input_value = cluster_number;
    var num_genes = $("#num-genes").val();
    var cell_color = $("#cell-color").val();
    console.log('update_barplot');
    var cluster1 = $("#barplot_cluster_select_1").val();
    var cluster2 = $("#barplot_cluster_select_2").val();
    var key = top_or_bulk + String(input_value) + ' ' + String(num_genes) + ' ' + String(cluster1) + ' ' + String(cluster2) + ' ' + String(cell_color);
    $("#update-area").empty();
    $("#update-area").append('Updating barplot <img src="/static/ajax-loader.gif"/>');
    if (cache.barplots.hasOwnProperty(key)) {
        var data = cache.barplots[key];
        var gene_names = data.data[0].y;
        $('#top-genes-view').val(gene_names.join('\n'));
        Plotly.newPlot("top-genes", data.data, data.layout);
        $("#update-area").empty();
        $("#update-area").append('Barplot updated');
        return true;
    }
    $.ajax({url: window.location.pathname + "/update_barplot",
        type: "POST",
        data: {"top_or_bulk": top_or_bulk,
               "input_value": input_value,
               "num_genes": num_genes,
               "cell_color": cell_color,
               "cluster1": cluster1,
               "cluster2": cluster2},
    }).done(function(data) {
        if (data.startsWith('Error')) {
            $("#update-area").empty();
            $("#update-area").append(data);
            return false;
        }
        data = JSON.parse(data);
        cache.barplots[key] = data;
        var gene_names = data.data[0].y;
        $('#top-genes-view').val(gene_names.join('\n'));
        if (gene_names.length > 20) {
            data.layout.yaxis = {};
            data.layout.yaxis.showticklabels = false;
        }
        data.data[0].x.reverse();
        data.data[0].y.reverse();
        Plotly.newPlot("top-genes", data.data, data.layout);
        $("#update-area").empty();
        $("#update-area").append('Barplot updated');
    });
    return true;
};

// this function is called on startup, and whenever the radio buttons
// corresponding to different input types are clicked.
function update_scatterplot() {
    var plot_type = $('input[name="scatter-type"]:checked').val();
    // TODO: have a cache and look up the outputs
    var cell_color = $("#cell-color").val();
    var gene_name = $('#gene_name').val();
    console.log(plot_type);
    console.log(cell_color);
    var gene_name_1 = $('#gene_name_1').val();
    var gene_name_2 = $('#gene_name_2').val();
    var use_mw = $('#use-mw').val();
    var cluster = $('#cluster_input').val();
    var upload_data = {"scatter_type": plot_type, "cell_color": cell_color,
               "gene_name": gene_name,
               "gene_name_1": gene_name_1,
               "gene_name_2": gene_name_2,
               "use_mw": use_mw,
               "cluster_input": cluster,
    };
    $("#update-area").empty();
    $("#update-area").append('Updating scatterplot <img src="/static/ajax-loader.gif"/>');
    var key = JSON.stringify(upload_data);
    // if the plot parameters have been used before, we retrieve them from the cache...
    if (cache.scatterplots.hasOwnProperty(key)) {
        var data = cache.scatterplots[key];
        Plotly.newPlot("means-scatter-plot", data.data, data.layout);
        current_scatterplot_data = data;
        bind_click();
        bind_select();
        $("#update-area").empty();
        $("#update-area").append('Scatterplot updated');
        return true;
    }
    $.ajax({url: window.location.pathname + "/update_scatterplot",
        type: "POST",
        data: upload_data,
    }).done(function(data) {
        if (data.startsWith('Error')) {
            $("#update-area").empty();
            $("#update-area").append(data);
            return false;
        }
        data = JSON.parse(data);
        cache.scatterplots[key] = data;
        Plotly.newPlot("means-scatter-plot", data.data, data.layout);
        current_scatterplot_data = data;
        bind_click();
        bind_select();
        $("#update-area").empty();
        $("#update-area").append('Scatterplot updated');
    });
    return true;
};

function set_enrichr_results(data, query) {
    // create table
    var results_view = $('#enrichr-results');
    if (query == 'cellmarker') {
        results_view = $('#cellmarker-results');
    }
    results_view.empty();
    var table = $('<table>');
    var header = $('<tr>');
    for (var j = 0; j<data[0].length; j++) {
        header.append('<th>'+data[0][j]+'</th>');
    }
    table.append(header);
    for (var i = 1; i<data.length; i++) {
        var row = $('<tr>');
        for (var j = 0; j<data[i].length; j++) {
            row.append('<td>'+data[i][j]+'</td>');
        }
        table.append(row);
    }
    results_view.append(table);
}

function update_gene_query(query) {
    // query is a string that can be 'enrichr', 'cellmarker', etc...
    // TODO: do something that makes this work for 
    var input_array = $('#gene-set-query-form').serializeArray();
    var data = {};
    $(input_array).each(function(index, obj){
        data[obj.name] = obj.value;
    });
    console.log('update_gene_query');
    console.log(data);
    var key = JSON.stringify(data);
    if (cache.enrichr.hasOwnProperty(key)) {
        set_enrichr_results(cache.enrichr[key], query);
        return true;
    }
    var results_view = $('#enrichr-results');
    var update_url = '/update_enrichr';
    if (query == 'cellmarker') {
        results_view = $('#cellmarker-results');
        update_url = '/update_cellmarker';
    }
    results_view.empty();
    results_view.append("<br>" + "Query in progress..." + '<img src="/static/ajax-loader.gif"/>');
    $.ajax({url: window.location.pathname + update_url,
        type: "POST",
        data: data,
    }).done(function(results) {
        results_view.empty();
        if (results.startsWith('Error')) {
            results_view.append(results);
        } else {
            results = JSON.parse(results);
            cache.enrichr[key] = results;
            set_enrichr_results(results, query);
        }
    });
    return true;
};

function bind_select() {
   var plot = $('#means-scatter-plot')[0];
   plot.on('plotly_selected', on_select);
};

// handler for scatterplot selection events
function on_select(data) {
    if (!data) {
        return false;
    }
    var clusters = new Set([]);
    // identify selected cells
    current_selected_cells = [];
    for (var i = 0; i<data.points.length; i++) {
        clusters.add(data.points[i].curveNumber);
        current_selected_cells.push(data.points[i].text);
    }
    // find cluster lengths
    var cluster_lengths = [];
    all_selected_clusters = [];
    var selection_string = "";
    clusters.forEach(function(x) {
        var l = current_scatterplot_data.data[x].x.length;
        cluster_lengths.push(l);
        all_selected_clusters.push(x);
        selection_string += "cluster " + String(x) + 
            " (" + String(l) + " cells), ";
    });
    // set update area...
    $("#update-area").empty();
    $("#update-area").append("selected clusters: " + selection_string + "<br>");
    $("#update-area").append("Number of selected cells: " + current_selected_cells.length);
};

// Gets basic cell info - read counts and gene counts
function get_cell_info() {
    var selected_clusters = all_selected_clusters;
    var selected_cells = current_selected_cells;
    // can't send an array... have to convert to string
    var upload_data = {
        selected_clusters: String(selected_clusters),
        selected_cells: String(selected_cells),
    };
    $("#update-area").empty();
    $("#update-area").append("<br>" + "Query in progress..." + '<img src="/static/ajax-loader.gif"/>');
    console.log(upload_data);
    $.ajax({url: window.location.pathname + "/cell_info",
        data: upload_data,
        method: 'POST',
    }).done(function(data) {
        var results = JSON.parse(data);
        var selection_string = "";
        selected_clusters.forEach(function(x) {
            var l = current_scatterplot_data.data[x].x.length;
            selection_string += "cluster " + String(x) + 
                " (" + String(l) + " cells), ";
        });
        var cell_string = "";
        for (var i = 0; i < selected_cells.length; i++) {
            cell_string += "cell " + String(selected_cells[i]) + " (" + String(results.gene_counts[i])
                +  " genes, " + String(results.read_counts[i]) + " reads), ";
        };
        $("#update-area").empty();
        $("#update-area").append("Selected cells: " + cell_string + "<br>");
        // TODO: what are the stats that we need?
        $("#update-area").append("Selected clusters: " + selection_string + "<br>");
    });
};

function split_or_merge_cluster(split_or_merge, cells_or_clusters) {
    if (currently_merging) {
        window.alert('Already running ' + split_or_merge + ' cluster');
        return false;
    }
    var result = window.confirm("Warning: splitting or merging might take a while, depending on the size of the dataset. Are you sure you wish to run this operation?");
    if (result == false) {
        return false;
    }
    var selected_clusters = all_selected_clusters;
    if (cells_or_clusters == "cells") {
        selected_clusters = current_selected_cells;
    }
    if (all_selected_clusters.length == 0) {
        window.alert('Warning: no selected clusters.');
        return false;
    }
    if (split_or_merge == "merge" && all_selected_clusters.length == 1) {
        window.alert('Warning: cannot merge a single cluster.');
        return false;
    }
    // lengths of each of the selected clusters
    console.log(split_or_merge);
    console.log(selected_clusters);
    $("#update-area").append("<br>" + split_or_merge + " clusters in progress... (re-running UNCURL, recalculating differentially expressed genes) " + '<img src="/static/ajax-loader.gif"/>');
    currently_merging = true;
    // make some indication that split/merge has been called.
    $.ajax({url: window.location.pathname + "/split_or_merge_cluster",
        data: {'split_or_merge': split_or_merge,
               'selected_clusters': selected_clusters.join(',')},
        method: 'POST',
    }).done(function(data) {
        currently_merging = false;
        if (data.startsWith('Error')) {
            $('#update-area').empty();
            $('#update-area').append(data);
        } else {
            $('#update-area').empty();
            $('#update-area').append(data);
            // reload page?
            cache.barplots = {};
            cache.scatterplots = {};
            update_scatterplot();
            update_barplot(0);
            //location.reload(true);
        }
    });
};

// copies data to a new user_id
function copy_data() {
    var result = window.confirm("Do you wish to copy this dataset to a new user id?");
    if (result == false) {
        return false;
    }
    $('#update-area').empty();
    $("#update-area").append("Copying data and results " + '<img src="/static/ajax-loader.gif"/>');
    $.ajax({url: window.location.pathname + "/copy_dataset",
        method: 'POST'
    }).done(function(data) {
        if (data.startsWith('Error')) {
            $('#update-area').empty();
            $('#update-area').append(data);
        } else {
            var new_user_id = data;
            $('#update-area').empty();
            $('#update-area').append('New results page: <a href="/user/' + new_user_id + '/view">' + new_user_id + '</a>');
        }
    });
};


// Starts a new uncurl sub-analysis...
function subset_cells(is_cells) {
    var result = window.confirm("Do you wish to run a new analysis on the selected cells?");
    if (result == false) {
        return false;
    }
    $('#update-area').empty();
    $("#update-area").append("Copying data, generating summary " + '<img src="/static/ajax-loader.gif"/>');
    var cell_ids = current_selected_cells;
    if (!is_cells) {
        cell_ids = all_selected_clusters;
    }
    console.log(is_cells);
    console.log(cell_ids);
    $.ajax({url: window.location.pathname + "/subset",
        data: {
            'is_cells': is_cells ? 1 : 0,
            'cell_ids': cell_ids.join(','),
        },
        method: 'POST'
    }).done(function(data) {
        if (data.startsWith('Error')) {
            $('#update-area').empty();
            $('#update-area').append(data);
        } else {
            var new_user_id = data;
            $('#update-area').empty();
            $('#update-area').append('New results page: <a href="/state_estimation/results/' + new_user_id + '">' + new_user_id + '</a>');
        }
    });
};

function toggle_reanalyze_area() {
    $('#reanalyze-area').toggle();
};

// toggles visibility of the bottom-left gene query view
// 'toggle' is a jquery method that changes the visibility of the element.
function toggle_query_visibility(value) {
    if (value == "cellmarker") {
        $('#cellmarker-view').toggle(true);
        $('#enrichr-view').toggle(false);
    } else if (value == "enrichr") {
        $('#cellmarker-view').toggle(false);
        $('#enrichr-view').toggle(true);
    }
};

window.onload = function() {
    // activate tooltips
    $('[data-toggle="tooltip"]').tooltip();
    // bind radio buttons to update_scatterplot
    $('input[name="scatter-type"]').change(function() {
        update_scatterplot();
    });
    $('#cell-color').change(function() {
        var cell_color = $("#cell-color").val();
        if (cell_color == "gene") {
            var gene_name = $('#gene_name').val();
            if (gene_name.length > 0) {
                update_scatterplot();
            }
            $('#gene-name-area').css('display', 'block');
            $('#color-track-upload-area').css('display', 'none');
            $('#cluster-select-area').css('display', 'none');
        } else if (cell_color == "new") {
            $('#color-track-upload-area').css('display', 'block');
            $('#gene-name-area').css('display', 'none');
            $('#cluster-select-area').css('display', 'none');
        } else if (cell_color == "weights") {
            $('#cluster-select-area').css('display', 'block');
            $('#gene-name-area').css('display', 'none');
            $('#color-track-upload-area').css('display', 'none');
        } else {
            $('#gene-name-area').css('display', 'none');
            $('#color-track-upload-area').css('display', 'none');
            $('#cluster-select-area').css('display', 'none');
            update_scatterplot();
        }
    });
    update_scatterplot();

    // bind dropdowns to update_barplot
    $('#top-or-bulk').change(function() {
        var option = $('#top-or-bulk').val();
        // if the selected option has "pairwise":
        if (option.includes('pairwise')) {
            $('#barplot-cluster-select-view').css('display', 'block');
            // set the select values for the current color scheme...
            var select_1 = $('#barplot_cluster_select_1');
            var select_2 = $('#barplot_cluster_select_2');
            // clear select children
            select_1.empty();
            select_2.empty();
            // get all cluster values
            for (var i in current_scatterplot_data.data) {
                var value = current_scatterplot_data.data[i];
                console.log('cluster ' + i);
                select_1.append($("<option>").attr('value', i).text(value.name));
                select_2.append($("<option>").attr('value', i).text(value.name));
            }
        } else {
            $('#barplot-cluster-select-view').css('display', 'none');
            update_barplot(currently_selected_cluster);
        }
    });
    $('#num-genes').change(function() {
        update_barplot(currently_selected_cluster);
    });
    update_barplot(currently_selected_cluster);

    // bind update enrichr button to update_enrichr
    $('#enrichr-submit').click(function() {
        update_gene_query('enrichr');
    });

    // bind merge and split buttons to split_or_merge_cluster
    $('#split').click(function() {
        split_or_merge_cluster('split', 'clusters');
    });
    $('#merge').click(function() {
        split_or_merge_cluster('merge', 'clusters');
    });
    $('#new-cluster').click(function() {
        split_or_merge_cluster('new', 'cells');
    });
    $('#copy').click(function() {
        copy_data();
    });

    $('#subset_cells').click(function() {
        // subset cells
        subset_cells(true);
    });

    $('#subset_clusters').click(function() {
        // subset clusters
        subset_cells(false);
    });
};

</script>


<style>

.wrapper {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    padding-top: 10px;
}

.top-left {
    grid-column: 1 / 2;
    grid-row: 1;
}

.top-right {
    grid-column: 2 / 3;
    grid-row: 1;
}

.bottom-left {
    grid-column: 1 / 3;
    grid-row: 2;
    margin-top: 20px;
}

.bottom-right {
    grid-column: 2 / 3;
    grid-row: 2;
    margin-top: 20px;
}

.container {
    padding-left: 0px;
}

table {
    padding-top: 5px;
    padding-bottom: 5px;
    padding-right: 5px;
}

tr {
}

td {
    padding-top: 10px;
    padding-right: 10px;
}

th {
    padding-top: 10px;
    padding-right: 10px;
}
</style>

{% endblock %}

{% block content %}

<div class="container container-main" role="main">
    <h3>Results for query id {{ user_id }}</h3>

    <p><a href="{{ url_for('data_stats', user_id=data_user_id) }}">Data stats</a></p>
    Results are available:
    {% block file_downloads %}
    <a href="{{ url_for('data_download', x=test_or_user, user_id=data_user_id) }}">Data downloads</a>

    {% endblock %}


    <div id="visualization" class="wrapper">

        <div class="top-left" id="left-hand-panel-1" style="width: 750px;">
            <!-- controls for which types of plots to show -->
            <div id="plot-type-options">
                <input type="radio" id="Means" value="Means" name="scatter-type" checked="checked" onclick="$('#gene-gene-names-area').css('display', 'none');">
                <label for="Means" data-toggle="tooltip" title="Scatterplot of the cluster archetypes, plotted by MDS.">Cluster means</label>
                <input type="radio" id="Cells" value="Cells" name="scatter-type" onclick="$('#gene-gene-names-area').css('display', 'none');">
                <label for="Cells" data-toggle="tooltip" title="Scatterplot visualization of cells after processing with UNCURL.">Post-processed cells</label>
                <input type="radio" id="Baseline" value="Baseline" name="scatter-type" onclick="$('#gene-gene-names-area').css('display', 'none');">
                <label for="Baseline" data-toggle="tooltip" title="Scatterplot visualization of cells before processing with UNCURL.">Pre-processed cells</label>
                <input type="radio" id="Gene-gene" value="Gene-gene" name="scatter-type" onclick="$('#gene-gene-names-area').css('display', 'block');">
                <label for="Gene-gene" data-toggle="tooltip" title="View expression levels of two genes.">Gene-gene</label>
            </div>


            <!-- controls for gene-gene plots -->
            <div id="gene-gene-names-area" style="display:none;" class="form-inline align-items-center">
                <div class="form-group">
                    <label for="gene_name_1">Gene 1:</label>
                    <input list="gene_names" id="gene_name_1" name="gene_name_1" class="form-control" style="width: 100px;">
                    <label for="gene_name_2">Gene 2:</label>
                    <input list="gene_names" id="gene_name_2"  name="gene_name_1" class="form-control" style="width: 100px;">
                </div>
                <div class="form-group" style="width:160px">
                    <select id="use-mw" class="form-control" style="width:150px">
                        <option value="0" selected>Unprocessed</option>
                <!-- TODO: "processed" won't work, because m only contains a subset of genes.
                        <option value="1">Processed</option>
                -->
                    </select>
                </div>
                <button class="btn btn-default" id="gene_gene_names_submit" onclick="update_scatterplot()">Submit</button>
            </div>

            <div id="means-scatter-plot" style="width: 700px; margin-top: 5px"></div>



            <div class="form-inline">
                <label for="cell-color" data-toggle="tooltip" title="How the cells should be colored - either by their cluster label, or by their cluster label entropy.">Label scheme:</label>
                <select class="form-control form-inline" id="cell-color" style="width: 300px;">
                    <option value="cluster" selected>Cluster</option>
                    <option value="entropy">Entropy</option>
                    <option value="gene">Gene</option>
                    <!-- TODO: add a view to show each cluster's W values --> 
                    <option value="weights">Cluster weights</option>
                    {% for track in color_tracks %}
                    <option value="{{ track }}">{{ track }}</option>
                    {% endfor %}
                    <option value="new">New color track</option>
                </select>
            </div>

            <div id="cluster-select-area" style="display:none;" class="form-inline align-items-center">
                <label for="cluster_input" data-toggle="tooltip" title="This shows the weights for one specified cluster.">Cluster ID:</label>
                <!-- TODO: only allow ints? -->
                <input type="number" min="0" step="1" id="cluster_input" name="cluster_input" class="form-control" style="width:200px;">
                <button type="button" class="btn btn-default" id="cluster_submit" onclick="update_scatterplot()">Submit</button>
            </div>

            <div id="gene-name-area" style="display:none;" class="form-inline align-items-center">
                <label for="gene_name">Gene name:</label>
                <input list="gene_names" id="gene_name" class="form-control" style="width: 200px;">
                <datalist id="gene_names">
                    {% for gene_name in gene_names %}
                    <option value="{{ gene_name }}">
                    {% endfor %}
                </datalist>
                <button class="btn btn-default" id="gene_name_submit" onclick="update_scatterplot()">Submit</button>
            </div>

            <div id="color-track-upload-area" style="display:none; width: 700px;">
                <form id="color-track-form" action="view/upload_color_track" method="post" enctype=multipart/form-data>
                    <div class="form-inline">
                        <label for="color_track_file" data-toggle="tooltip" title="File should contain one line for each cell.">Upload color track here</label>
                        <input type="file" name="color_track_file" id="color_track_file" class="form-control-file">
                        <label for="color_track_type" data-toggle="tooltip" title="Discrete: each line will be interpreted as a string/label; continuous: each line will be interpreted as a number; table: TODO">Type of color track:</label>
                        <select class="form-control" id="color_track_type" name="color_track_type" style="width: 200px;">
                            <option value="continuous" selected>Continuous</option>
                            <option value="discrete">Discrete</option>
                            <option value="table">Table</option>
                        </select>
                        <button class="btn btn-default" type="submit" id="submit_color_track" class="btn btn-default">Submit</button>
                    </div>
                </form>
            </div>

            <div class="input-group" style="margin-top: 5px;">
                <div class="form-row" id="update-area"></div>

                <button type="button" class="btn btn-default" id="cell-info" onclick="get_cell_info();">Get cell/cluster info</button>
                <button type="button" class="btn btn-default" id="split" style="margin-right: 2px;" data-toggle="tooltip" title="Split the selected cluster into two clusters">Split clusters</button>
                <button type="button" class="btn btn-default" id="merge" data-toggle="tooltip" title="Merge the selected clusters into one cluster">Merge clusters</button>
                <button type="button" class="btn btn-default" id="new-cluster" data-toggle="tooltip" title="Create a new cluster from the selected clusters">Create new cluster</button>
                <button type="button" class="btn btn-default" id="reanalyze" onclick="toggle_reanalyze_area();">Reanalyze</button>

                <div id="reanalyze-area" style="display:none;">
                    <button class="btn btn-default" id="copy">Copy dataset</button>
                    <button class="btn btn-default" id="subset_cells">Sub-select cells</button>
                    <button class="btn btn-default" id="subset_clusters">Sub-select clusters</button>
                </div>
            </div>
        </div>

        <div class="top-right" id="right-hand-panel-1" style="width: 550px">
            <select class="form-control" id="top-or-bulk" style="width: 300px;" data-toggle="tooltip" title="Which quantity to visualize in the barplot. C-score is the ratio of the mean expression in the given cluster to the cluster with the second-highest expression.">
                <option value="top">Top genes (c-score)</option>
                <option value="pval">Top genes (p-value of c-score)</option>
                <option value="top_1_vs_rest" selected>Top genes (1-vs-rest fold change)</option>
                <option value="pval_1_vs_rest">Top genes (p-value of fold change)</option>
                <option value="selected_color">Top genes for custom labels (1 vs rest fold change)</option>
                <!--<option value="selected_color_pval">Top genes for custom labels (1 vs rest pval)</option>-->
                <option value="sep">Most similar clusters (separation score)</option>
                <option value="top_pairwise">Pairwise fold change</option>
                <option value="pval_pairwise">Pairwise p-value</option>
                <!--<option value="bulk">Bulk correlations</option>-->
            </select>

            <div style="display:none;" id="barplot-cluster-select-view" class="form-inline align-items-center">
                <!-- this contains an interface to select a cluster to do pairwise diffexp against. -->
                <!-- This can either be clusters or custom labels -->
                <label for="barplot_cluster_select_1">Cluster 1:</label>
                <select class="form-inline form-control" id="barplot_cluster_select_1" name="barplot_cluster_select_1" onchange="update_barplot(0);">
                </select>
                <label for="barplot_cluster_select_2">Cluster 2:</label>
                <select class="form-inline form-control" id="barplot_cluster_select_2" name="barplot_cluster_select_2" onchange="update_barplot(0);">
                </select>
            </div>

            <div style="display:none;" id="barplot_gene_select_view" class="form-inline align-items-center">
                <!-- TODO: this contains an interface to select a gene for comparing the selected cluster vs all. -->
                <label for="barplot_gene_select">Gene:</label>
                <select class="form-inline form-control" id="barplot_gene_input" name="barplot_gene_input">
                </select>
            </div>

            <div style="margin-top: 4; margin-bottom: 0; width: 300px;" class="form-inline align-items-center">
                <label for="num-genes">Number of top genes:</label>
                <input type="number" class="form-inline form-control" id="num-genes" value="10">
            </div>

            <!-- Plot of top gene/p-value bar plots goes here -->
            <div id="top-genes" style="width: 500px;"></div>

        </div>

        <!-- Bottom left panel - for gene set queries on databases. "bottom-left" actually covers the entire bottom row, since there's no bottom right. -->
        <div class="bottom-left" id="bottom-panel" style="margin-top: 20px; width: 1000px;">
            <h4>Gene set queries</h4>
            <form id="gene-set-query-form">
                <div>Currently selected top genes:</div>
                <div class="form-inline">
                    <textarea class="form-control" id="top-genes-view" cols="50" rows="4" name="top_genes">Top genes go here</textarea>
                </div>
                <!-- onchange: make the cellmarker view visible -->
                <div class="form-inline">
                    <label for="database-select" data-toggle="tooltip" title="Which database to query for the given gene set. Current options are Enrichr and CellMarker.">Database:</label>
                    <select class="form-control" id="database-select" name="database_select" style="width:200px; margin-top: 5px;" onchange="toggle_query_visibility(this.value);">
                        <option value="enrichr" selected>Enrichr</option>
                        <option value="cellmarker">CellMarker</option>
                    </select>
                </div>

                <div id="enrichr-view">
                    <h5 data-toggle="tooltip" title="Uses the Enrichr tool from http://amp.pharm.mssm.edu/Enrichr/">Call Enrichr on selected genes</h4>
                    <div style="margin-top: 5px; margin-bottom: 5px; width: 300px;" class="form-inline">
                        <label for="gene-set-library">Gene set:</label>
                        <select class="form-control" id="gene-set-library" name="gene_set">
                            {% for opt in gene_sets %}
                            <option value={{ opt }}>{{ opt }}</option>
                            {% endfor %}
                        </select>
                    </div>
                    <button class="btn btn-default" id="enrichr-submit" type="button">Submit Enrichr query</button>
                    <div id="enrichr-results"></div>
                </div>

                <div id="cellmarker-view" style="display:none;">
                    <h5 data-toggle="tooltip" title="Query the CellMarker database for cell types matching the selected genes.">Query CellMarker with selected genes</h4>
                    <div class="form-inline" style="margin-top: 5px; margin-bottom: 5px; width: 300px;">
                        <label for="test-type" data-toggle="tooltip" title="Currently, only the hypergeometric test is available.">Statistical test:</label>
                        <select class="form-control" id="test-type" name="test_type">
                            <option value="hypergeom" selected>Hypergeometric test</option>
                        </select>
                    </div>

                    <div class="form-inline" style="margin-top: 5px; margin-bottom: 5px; width: 300px;">
                        <label for="species" data-toggle="tooltip" title="Which species to query for cell types. Currently human and mouse are available.">Species:</label>
                        <select class="form-control" id="species" name="species">
                            <option value="all" selected>All</option>
                            <option value="Human">Human</option>
                            <option value="Mouse">Mouse</option>
                        </select>
                    </div>

                    <div class="form-inline" style="margin-top: 5px; margin-bottom: 5px; width: 300px;">
                        <label for="cells-or-tissues">Cells or tissues:</label>
                        <select class="form-control" id="cells-or-tissues" name="cells_or_tissues">
                            <option value="cells" selected>Cells</option>
                            <option value="tissues">Tissues</option>
                        </select>
                    </div>

                    <button class="btn btn-default" id="cellmarker-submit" type="button" onclick="update_gene_query('cellmarker')">Submit CellMarker query</button>
                    <div id="cellmarker-results"></div>
                </div>
            </form>
        </div>

    </div>

    <p>Results</p>

    <!-- TODO: display properties in this window... -->
    <div id="data_properties"></div>


</div>

{% endblock %}
